File: MAX2870.cpp
#include <SPI.h>
#include "MAX2870.h"
#include <math.h>
#include <stdio.h>

//****************************************************************************
MAX2870::MAX2870(const uint8_t MAX2870_pin_LE, const uint8_t MAX2870_pin_CE, const uint8_t MAX2870_pin_LD) {
  delay(20);

  pin_LE = MAX2870_pin_LE;
  pin_CE = MAX2870_pin_CE;
  pin_LD = MAX2870_pin_LD;

  pinMode (pin_LE, OUTPUT);
  digitalWrite(pin_LE, 0);

  pinMode (pin_CE, OUTPUT);
  digitalWrite(pin_CE, 1);

  pinMode (pin_LD, INPUT);

  SPI.setClockDivider(SPI_CLOCK_DIV16); //16MHz system clock \ 16 = 1MHz SPI
  SPI.setDataMode(SPI_MODE0); //CPOL = CPHA = 0, 8 bits per frame
  SPI.setBitOrder(MSBFIRST);
  SPI.begin();

  delay(20);

  reg0.all = 0x007d0000;
  reg1.all = 0x2000fff9;
  reg2.all = 0x00004042;
  reg3.all = 0x0000000b;
  reg4.all = 0x6180b23c;
  reg5.all = 0x00400005;
  reg6.all = 0x00000000;

  setConfig(20);

  delay(20);

  setConfig();
}

void MAX2870::pulseLE() {
  digitalWrite(pin_LE, HIGH);
  delayMicroseconds(1);
  digitalWrite(pin_LE, LOW);
  delayMicroseconds(1);
}

//****************************************************************************
void MAX2870::writeData(uint32_t data) {

  SPI.transfer((0xFF000000 & data) >> 24);  
  SPI.transfer((0x00FF0000 & data) >> 16);
  SPI.transfer((0x0000FF00 & data) >> 8);
  SPI.transfer( 0x000000FF & data);

  pulseLE();

}

//****************************************************************************
void MAX2870::setConfig(uint8_t first_delay) {
  writeData(reg5.all);
  if (first_delay) delayMicroseconds(first_delay);
  writeData(reg4.all);
  writeData(reg3.all);
  writeData(reg2.all);
  writeData(reg1.all);
  writeData(reg0.all);
}

void MAX2870::setConfig() {
  setConfig(0);
}

//****************************************************************************
void MAX2870::set_RF_OUT_A(double freq) {
  uint32_t n, frac, m, diva = 0;
  double pll_coefficient, fractional = 0;

  while (freq * powf(2, diva) < 3000.0)  {
    diva = diva + 1;
  }
  pll_coefficient = freq * powf(2, diva) / f_pfd;
  n = floor(pll_coefficient);

  fractional = pll_coefficient - n;
  m = 4000;
  frac = round(m * fractional);

  reg0.bits.frac = frac;
  reg0.bits.n = n;
  reg1.bits.m = m;
  reg4.bits.diva = diva;


  setConfig();
  f_rfouta = f_pfd * (reg0.bits.n + 1.0 * reg0.bits.frac / reg1.bits.m) / powf(2, reg4.bits.diva);
}

void MAX2870::setPFD(const double ref_in, const uint16_t rdiv) {
  // fPFD = fREF * [(1 + DBR)/(R * (1 + RDIV2))]  
  // DBR=0 RDIV2=0
  // fPFD = fREF * [1/R] = fREF / R
  //
  
  f_pfd = ref_in / rdiv;

  if (f_pfd > 32.0) {
    reg2.bits.lds = 1;
  }
  else {
    reg2.bits.lds = 0;
  }

  reg3.bits.cdiv = round(f_pfd / 0.10);

  reg2.bits.dbr = 0;
  reg2.bits.rdiv2 = 0;
  reg2.bits.r = rdiv;

  uint32_t bs = f_pfd * 20;

  if (bs > 1023) {
    bs = 1023;
  }
  else if (bs < 1) {
    bs = 1;
  }

  reg4.bits.bs = 0x03FF & bs;
  reg4.bits.bs2 = 0x03 & (bs >> 8);

  setConfig();
}

void MAX2870::dump_regs(uint32_t* dump) {
  dump[0] = reg0.all;
  dump[1] = reg1.all;
  dump[2] = reg2.all;
  dump[3] = reg3.all;
  dump[4] = reg4.all;
  dump[5] = reg5.all;
  dump[6] = reg6.all;
  return;
}

void MAX2870::powerOn(bool pwr) {
  reg2.bits.shdn =  !pwr;

  setConfig();
}

void MAX2870::aout(uint8_t pwr) {
  reg4.bits.rfa_en = pwr ? 1 : 0;
  reg4.bits.apwr = pwr - 1;

  setConfig();
}

void MAX2870::bout(uint8_t pwr, bool bdiv = 0) {
  reg4.bits.rfb_en = pwr ? 1 : 0;
  reg4.bits.bpwr = pwr - 1;
  reg4.bits.bdiv = bdiv;

  setConfig();
}

uint8_t MAX2870::get_ld()
{
  uint8_t res;
  res = digitalRead(pin_LD);
  return res;
}

File: max2870.py
import wiringpi
from wiringpi import GPIO
import time


class max2870:
	
	init_tx = [
		0x00400005, 
		0x00000004, 
		0x00000003, 
		0x00000002, 
		0x00000001, 
		0x00000000 
	]	
	
	def __init__(self, ce_pin, le_pin, F_REF, MOD = 4000, R_DIV = 2):
		self.ce_pin = ce_pin
		self.le_pin = le_pin
		self.F_REF = F_REF
		self.M = MOD
		self.R = R_DIV
		
		
		wiringpi.wiringPiSPISetupMode(0, 0, 500000, 0)

		wiringpi.wiringPiSetup()

		wiringpi.pinMode(self.ce_pin, GPIO.OUTPUT)
		wiringpi.pinMode(self.le_pin, GPIO.OUTPUT)

		wiringpi.digitalWrite(self.ce_pin, GPIO.HIGH)
		wiringpi.digitalWrite(self.le_pin, GPIO.LOW)

		self.write_reg(self.init_tx[0])
		time.sleep(0.020)
		self.write_reg(self.init_tx[1])
		self.write_reg(self.init_tx[2])
		self.write_reg(self.init_tx[3])
		self.write_reg(self.init_tx[4])
		self.write_reg(self.init_tx[5])

		time.sleep(0.1)
		self.write_reg(self.init_tx[0])
		self.write_reg(self.init_tx[1])
		self.write_reg(self.init_tx[2])
		self.write_reg(self.init_tx[3])
		self.write_reg(self.init_tx[4])
		self.write_reg(self.init_tx[5])
		
		self.calc_f_pfd()
		
		self.P      = 1
		self.cpoc   = 0
		self.cpt    = 0
		self.cpl    = 0
		
		self.rdiv2  = 0
		self.sdn    = 0
		self.mux    = 0
		self.dbr    = 0
		self.pdp    = 1
		self.rst    = 0
		self.tri    = 0
		self.shdn   = 0
		self.tri    = 0
		self.ldp    = 0
		self.ldf    = 0
		self.cp     = 0
		self.reg4p  = 0
		self.vas_shdn = 0
		self.vco    = 0
		self.retune = 0
		self.cdm    = 0
		self.int    = 0
		self.bdiv   = 0
		self.bpwr   = 0
		self.rfb_en = 0
		self.apwr   = 3
		self.rfa_en = 1
		self.fb     = 1 #N counter mode = fundamental

	def pulse_le(self):
		wiringpi.digitalWrite(self.le_pin, GPIO.HIGH)
		time.sleep(0.0001)
		wiringpi.digitalWrite(self.le_pin, GPIO.LOW)
		time.sleep(0.0001)
	
	def write_reg(self, x):
		r = [(x & 0xff000000) >> 24,
			 (x & 0x00ff0000) >> 16,
			 (x & 0x0000ff00) >>  8,
			 (x & 0x000000ff)]
		revlen, recvData = wiringpi.wiringPiSPIDataRW(0, bytes(r))
		self.pulse_le()

	def get_diva(self, freq_kHz):
		self.diva = 0
		while (freq_kHz * 2**self.diva < 3000000):
			self.diva = self.diva + 1;
		
		return self.diva
		
	def calc_f_pfd(self):
		self.f_pfd = self.F_REF / self.R;

		self.bs = int(self.f_pfd / 50000);
		if (self.bs > 1023):
			self.bs = 1023;
	
		self.cdiv = int(self.f_pfd / 100000);

		self.lds = 0;
		if (self.f_pfd > 32000000) :
			self.lds = 1;

	def prep_regs(self):
		# --- Формуємо регістри ---
		self.R0 = (self.int  << 31) | \
				  (self.N    << 15) | \
				  (self.FRAC <<  3) | \
				  0;
				  
		self.R1 = (self.cpoc  << 31) | \
				  (self.cpl   << 29) | \
				  (self.cpt   << 27) | \
				  (self.P     << 15) | \
				  (self.M     <<  3) | \
				  1;

		# R2 – включає PFD, R divider
		self.R2 = (self.lds   << 31) | \
				  (self.sdn   << 29) | \
				  (self.mux   << 26) | \
				  (self.dbr   << 25) | \
				  (self.rdiv2 << 24) | \
				  (self.R     << 14) | \
				  (self.reg4p << 13) | \
				  (self.cp    <<  9) | \
				  (self.ldf   <<  8) | \
				  (self.ldp   <<  7) | \
				  (self.pdp   <<  6) | \
				  (self.shdn  <<  5) | \
				  (self.tri   <<  4) | \
				  (self.rst   <<  3) | \
				  2;

		self.R3 = (self.vco      << 26) | \
				  (self.vas_shdn << 25) | \
				  (self.retune   << 24) | \
				  (self.cdm      << 15) | \
				  (self.cdiv     <<  3) | \
				  3;  

		self.R4 = (0x18 << 26) | \
				  (((self.bs >>  8) & 0x03) << 24) | \
				  (self.fb        << 23) | \
				  (self.diva      << 20) | \
				  ((self.bs & 0xff) << 12) | \
				  (self.bdiv      <<  9) | \
				  (self.rfb_en    <<  8) | \
				  (self.bpwr      <<  6) | \
				  (self.rfa_en    <<  5) | \
				  (self.apwr      <<  3) | \
				  4
				   
		self.R5 = 0x400005;  # стандартний

	def setFrequency(self, freq_kHz):

		if (freq_kHz < 23500): return 0
		if (freq_kHz > 6000000): return 0 

		self.get_diva(freq_kHz)

		ratio = freq_kHz * 1000 * 2**self.diva / self.f_pfd
		self.N = int(ratio);
		self.FRAC = int((ratio - self.N) * self.M);
		#print(f"f_pfd:{self.f_pfd}; diva:{diva}; ratio:{ratio}; INT:{self.INT}; FRAC:{self.FRAC}; bs:{bs}; cdiv:{cdiv}");
	
		self.prep_regs()

		# --- Відправляємо регістри ---
		self.write_reg(self.R5)
		self.write_reg(self.R4)
		self.write_reg(self.R3)
		self.write_reg(self.R2)
		self.write_reg(self.R1)
		self.write_reg(self.R0)

		freq_res = self.f_pfd * (self.N + self.FRAC / self.M) / 2**self.diva
		
		return freq_res / 1000

	
	def __str__(self):
		return f"{self.R0:08X}\n{self.R1:08X}\n{self.R2:08X}\n{self.R3:08X}\n{self.R4:08X}\n{self.R5:08X}\n";
	




